common-http/src/main/java/io/cdap/common/http/HttpResponse.java

```diff
@@ -54,21 +75,44 @@ public String getResponseMessage() {
   }
 
   public byte[] getResponseBody() {
+    if (responseBody == null) {
+      responseBody = getResponseBodyFromStream();
+    }
     return responseBody;
   }
 
   public String getResponseBodyAsString() {
-    return new String(responseBody, Charsets.UTF_8);
+    return new String(getResponseBody(), Charsets.UTF_8);
   }
 
   public String getResponseBodyAsString(Charset charset) {
-    return new String(responseBody, charset);
+    return new String(getResponseBody(), charset);
   }
 
   public Multimap<String, String> getHeaders() {
     return headers;
   }
 
+  public void consumeContent(HttpResponseConsumer consumer) throws IOException {
+    if (inputStream == null) {
+      consumer.onFinished();
+      return;
+    }
+
+    try (ReadableByteChannel channel = Channels.newChannel(inputStream)) {
+      ByteBuffer buffer = ByteBuffer.allocate(consumer.getChunkSize());
+      while (channel.read(buffer) > 0) {
```

Comment:
Should be >= 0. It's possible for channel.read to return 0, which doesn't signal the end of stream

Reply:
I see, thanks for pointing that out! Fixed.